website/docs/documentation/helpers/only.mdx

```diff
@@ -0,0 +1,171 @@
+---
+sidebar_position: 11
+tags: [modifiers, debugging]
+---
+
+import { History } from '@site/src/components/History';
+import { Stability } from '@site/src/components/Stability';
+
+# 🌌 only
+
+`.only` is an extended helper for `describe`, `it`, and `test` to assist you debug tests.
+
+<Stability
+  level={1}
+  message={
+    <>
+      <div>
+        <strong>Poku</strong> doesn't map tests to run them, allowing native{' '}
+        <strong>JavaScript</strong> syntax in tests.
+      </div>
+      <div>
+        As a result, a different approach was adopted to allow{' '}
+        <code>.only</code> runs.
+      </div>
+    </>
+  }
+/>
+
+<History
+  records={[
+    {
+      version: '2.7.0',
+      changes: [
+        <>
+          Add <code>only</code> modifier to <code>describe</code>,{' '}
+          <code>it</code> and <code>test</code> methods.
+        </>,
+      ],
+    },
+  ]}
+/>
+
+## Usage
+
+You need to pass one of the temporary flags below to use the `.only` helper:
+
+> Look for "👋🏻" in the following examples.
+
+### `--only`
+
+- Only tests using `.only` will be run. This applies to `describe`, `it` and `test` at the same time.
+- If a common `describe` has a `it.only` in its scope, for example, the `it.only` will not be reached (see [**`--itOnly`**](#--itonly)).
```

Comment:
Below it is documented that `--only` is the same as `--describeOnly --itOnly`. Hm.. Looks like these are contradicting statements. Or?

Reply:
Also, did you consider: `--only=it`, `--only=test`, `--only=describe,test`? The jump from `--only` to `--describeOnly` does not feel intuitive. If I want the `only` feature, makes me think of typing `--only` first and optionally adding  something to it. Or?

Reply:
Another aspect, what happens with the after/before hooks? And is the code contain within `describe()` executed at some point? Or how? Not clear.

Reply:
> `--only=it`, `--only=test`, `--only=describe,test`

That sounds interesting.

---

> Below it is documented that `--only` is the same as `--describeOnly --itOnly`. Hm.. Looks like these are contradicting statements. Or?

I'm not sure I understand 😅

---

> Another aspect, what happens with the after/before hooks? And is the code contain within `describe()` executed at some point? Or how? Not clear.

Everything should work exactly the same with or without the `.only`. The `.only` just calls the base `it` or `describe` methods, for example:

https://github.com/wellwelwel/poku/blob/8dcce80c7adefecc096ca235bf634e437053f020/src/modules/helpers/modifiers.ts#L87-L88

Reply:
> `--only=it`, `--only=test`, `--only=describe,test`

I really appreciate this approach. I think it's worth removing the "experimental". Thanks, @mrazauskas.

I'll add to the documentation that everything should work equally in tests that use `.only`.

About the "contradicting", could you elaborate it? 🙋🏻‍♂️

Reply:
Please, understand me right (; I just try to figure out why it was not enough to have `---only`? Why `--describeOnly` and `--testOnly` are needed? Perhaps the idea was that `--describeOnly` does not execute `describe()`, but only `describe.only()` get executed? In the other hand, it feels like `--testOnly` would executed all `describe()` code, because they all get visited to find `test.only()`. What `--only` does, currently is unclear to me.

Reply:
Regarding the contradiction. First statement: "If a common `describe` has a `it.only` in its scope, the `it.only` will not be reached". Second statement "`npx poku --only # Same as: npx poku --describeOnly --itOnly`".

`--itOnly` does not reach `it.only` inside a `describe` without `only`? Hm.. Felt like that’s the purpose of this flag.

In other words, if `--only` does not reach `it.only` inside a `describe` without `only`, it means that `--only` is not the same as `--describeOnly --itOnly`. Or?

Reply:
The `--only` forces the user to set the `.only` in both `describe` and `it`/`test`, for example:

```ts
describe('it will be ignored', () => {
  test.only("since this `test` is in a context/scope that will never be called, it will never be reached.")
});
```

To "fix" this, the user needs to use `.only` in both:

```ts
describe.only('it will be executed', () => {
  test.only("it will be executed")
});
```

The `--itOnly` allows to use `describe` without `.only` and will execute only `it.only`/`test.only` tests (inside a `describe` or not).

Reply:
> In other words, if `--only` does not reach `it.only` inside a `describe` without `only`, it means that `--only` is not the same as `--describeOnly --itOnly`. Or?

If you use both like `poku --itOnly --describeOnly`, it will require that both uses `.only`:

```ts
describe.only('it will be executed', () => {
  test.only("it will be executed", () => {
  });
});
```

Reply:
Thanks for explanation. So it feel like this is not 100% correct: `npx poku --only # Same as: npx poku --describeOnly --itOnly`. Or?

Reply:
Here:

https://github.com/wellwelwel/poku/blob/8dcce80c7adefecc096ca235bf634e437053f020/website/docs/documentation/helpers/only.mdx?plain=1#L88-L93

Reply:
> So it feel like this is not 100% correct: `npx poku --only # Same as: npx poku --describeOnly --itOnly`. Or?

If there's a difference, it's not what is expected.

For example, they're parsed separately:

https://github.com/wellwelwel/poku/blob/8dcce80c7adefecc096ca235bf634e437053f020/src/parsers/get-arg.ts#L64-L68

But, in practice:

https://github.com/wellwelwel/poku/blob/8dcce80c7adefecc096ca235bf634e437053f020/src/modules/helpers/modifiers.ts#L78

https://github.com/wellwelwel/poku/blob/8dcce80c7adefecc096ca235bf634e437053f020/src/modules/helpers/modifiers.ts#L53

Reply:
Perhaps an approach documenting this as _"to enable `.only` in `it` tests ..."_ and _"to enable `.only` in `describe` tests ..."_ can be a more effective way.

Reply:
Not sure I follow. Hm..

> If there's a difference, it's not what is expected.

`--only` does not include `test.only` from within `describe` that has no only. `--testOnly` does include all and every `test.only`. How `--only` and `--describeOnly --itOnly` can be the same?

It feels like `--only` includes only the top level `.only`s. But `--describeOnly` or `--itOnly` also reach the nested ones. Or I missed something?

Reply:
About `--testOnly`, the `test` method is just an alias for the `it` method:

https://github.com/wellwelwel/poku/blob/8dcce80c7adefecc096ca235bf634e437053f020/src/modules/helpers/test.ts#L1-L3

Everything done in `it` will automatically be reflected in `test`. In other words, `testOnly` is the same as `itOnly`:

https://github.com/wellwelwel/poku/blob/8dcce80c7adefecc096ca235bf634e437053f020/src/parsers/get-arg.ts#L66

Reply:
Thanks for code examples, but all I want is to understand the intended behaviour. Could you answer if:

- `--only` and `--describeOnly --itOnly` are the equivalent? Or not?
- `--only` does include `test.only` from within `describe` that has no `only`, or not?
- `--describeOnly ` does include `test.only` from within `describe` that has no `only`, or not?

Reply:
> It feels like `--only` includes only the top level `.only`s. But `--describeOnly` or `--itOnly` also reach the nested ones. Or I missed something?

The `.only` doesn't distinguish scope. The reason `it.only` isn't executed if it's in a _normal_ `describe` when using `--only`, is that only tests with `.only` will be executed:

> The "normal" `describe`:

https://github.com/wellwelwel/poku/blob/8dcce80c7adefecc096ca235bf634e437053f020/src/modules/helpers/describe.ts#L80

This means that **JavaScript** will never reach `it.only` inside a `describe` without `.only`.

Reply:
> * `--only` and `--describeOnly --itOnly` are the equivalent? Or not?

Yep, they're equivalent 🙋🏻‍♂️

> * `--only` does include `test.only` from within `describe` that has no `only`, or not?

Since `test` is an alias for `it`, `--testOnly` is also an alias for `itOnly`, so `--only` applies to `describe`, `it` and `test` methods, while `itOnly` or `testOnly` applies for both `test` and `it` methods.

> * `--describeOnly ` does include `test.only` from within `describe` that has no `only`, or not?

Yes, only `describe.only` will be affected by `--describeOnly`. If a `test`, `it`, `test.only` or `it.only` is inside a _normal_ `describe` when using `describeOnly`, they won't be reached. If they (`it` and `test`) are _top-level_, they will just follow the standard JavaScript flow.

Reply:
If `--only` and `--describeOnly --itOnly` are equivalent, the lines to which I added the original comment are incorrect or at least unclear.

Reply:
> If `--only` and `--describeOnly --itOnly` are equivalent, the lines to which I added the original comment are incorrect or at least unclear.

Maybe the ``"Only tests using `.only` will be run."``, is what's causing the confusion, since I also considered `describe` as a type of test.

I'll take a look at this approach:

> `--only=it`, `--only=test`, `--only=describe,test`

And then better elaborate the documentation and the choice of words.

Reply:
Thus sounds strange: "the `it.only` will not be reached". Hm.. Does "reached" here mean "executed"? But hm.. everything with `.only` gets executed. Or?

Reply:
By "unreached", I mean the JavaScript syntax (because the tests are not mapped).

“Not executed” fits for `skip` approach, for example.

It's like:

```ts
const cond = true;

describe(() => {
  if (cond) return;

  // this will never be reached
  it(() => true);
});
```

Which is similar to what happens here (line 80):

https://github.com/wellwelwel/poku/blob/8dcce80c7adefecc096ca235bf634e437053f020/src/modules/helpers/describe.ts#L80-L83

- The line 82 will never be reached if `--only` or `--describeOnly` is used.

Reply:
> This sounds strange: "the `it.only` will not be reached". Hm.. Does "reached" here mean "executed"? But hm.. everything with `.only` gets executed. Or?

`describe` will not be executed, because of this, everything within its scope won't be either.

Reply:
Right. `describe` is not executed because of missing `only`. But `--only` and `--onlyIt --onlyDescribe` are equivalent.

So can they be equivalent if `--only` does not reach what `--onlyIt` does reach?

Sorry, Sir. We are in a loop (;

Reply:
My approach was to try to explain implicitly that not mapping the tests brings different behaviors from the most popular test runners.

I believe that the ideal is to focus on what **Poku** does and not on what the others do or don't do.

When tests are mapped, it makes this method easier to handle on the development side, so trying to bring in the costs of not mapping the tests seems to be overcomplicating things.

Reply:
> I've just noticed that I don't know how to laugh in English 🧐

Reply:
@mrazauskas, I released the `2.7.0` with this usage:

- `--only`
- `--only=it`
- `--only=test`
- `--only=describe`

About the documentation, it would be great if you could check it out: https://poku.io/docs/documentation/helpers/only

I tried to divide it into sections, starting with simpler cases and including more complex ones as the section progresses.

Reply:
Thanks for the update. I read it carefully twice and I still have many questions. Uff.. I will play with it first and come back with feedback at some point.

---

As a quick examples:

```js
describe.only(() => {
  test("one", () => {
    // ...
  });
});

test("two", () => {
  // ...
});
```

Description of `--only=describe` says: "`it` and `test` methods without `.only` will be executed". All of them? I mean, `"two"` will be executed as well? That sounds unexpected. I have to try.

---

<img width="631" alt="Screenshot 2024-09-17 at 09 40 01" src="https://github.com/user-attachments/assets/19cb998c-218d-452a-8e51-e3a3a495a179">

Hm.. Because it is commented out? Or? (;

---

<img width="507" alt="Screenshot 2024-09-17 at 09 46 23" src="https://github.com/user-attachments/assets/d357538c-3bbe-4a49-b8ff-8948be15228e">

Before / after **all** or before / after **each**?

Reply:
One more: behaviour example of `.only` within nested `describe` blocks is missing. `describe` > `describe.only` > `test`. Is this `describe.only` executed with `--only=describe`? Or the parent `describe` gets skipped?

Reply:
> Description of `--only=describe` says: "`it` and `test` methods without `.only` will be executed". All of them? I mean, `"two"` will be executed as well? That sounds unexpected. I have to try.

This one is more of a limitation than an intention. To ensure that `test` and `it` outside the scope of `describe` method are also ignored, the ideal is to use `--only`.

> Hm.. Because it is commented out? Or? (;

😂

> Before / after **all** or before / after **each**?

Ops, that should be "each".

Reply:
> One more: behaviour example of `.only` within nested `describe` blocks is missing. `describe` > `describe.only` > `test`. Is this `describe.only` executed with `--only=describe`? Or the parent `describe` gets skipped?

If the first `describe` is skipped, everything within its scope/context will be skipped too 🙋🏻‍♂️

Reply:
Trying it out:

```js
import { describe, test } from "poku";

describe(() => {
  describe.only(() => {
    test("one", () => {
      //
    });
  });
});

test("two", () => {
  //
});
```

- `--only=describe` runs `"two"`. The command is about `only` and `describe`, but it runs a test that does not have `only` and is not within a `describe`. Puzzling? Isn’t it?

---

```js
import { describe, test } from "poku";

describe.only(() => {
  test("one", () => {
    //
  });
});

test("two", () => {
  //
});
```

- `--only` runs nothing. Hm..

Reply:
Ah.. I see. Second example is my mistake. To make it work as I expect, if this pattern would be implemented:

```js
import { test } from "poku";

test.only((t) => {
  t.test("one", () => {
    //
  });
});

test("two", () => {
  //
});
```

Reply:
> * Puzzling? Isn’t it?

At this point, it's important to consider that `--only=describe` is different from "mapping the `describe` methods", but rather "`enabling `.only` for the `describe` methods`". To use `.only` for `test` and `it`, it's necessary to enable it for them.

> * Now I think this is by design

Yes, the `describe` is executed, but the "one" doesn't, since it doesn't use the `.only` with `--only`.

Reply:
The limitations of not mapping the tests to the `.only` case while trying to bring about a similar behavior to the usual one were holding me back both technically and creatively.

The idea of switching on and off and the concept of “enabling”, helped me to reconcile a different approach without losing the idea of **JavaScript** syntax.